### ISSUE TITLE
CSP script-src self and blobs

### ISSUE THREAD

**Issue by youennf** [w3c/webappsec-csp#487]
When looking at blob loads with CSP, I noted a few things:
- frame-src 'self' will allow blob navigations and downloads.
- script-src 'self' will make blob script loads fail.

I think it is ok to be tighter for blob URLS in case of script but I do not find these rules in the spec.
But from looking at the spec, I do not see where script-src differs from frame-src on that respect.
They seem to both end up executing https://w3c.github.io/webappsec-csp/#match-url-to-source-expression, which might return true as per step 4.1.

It might be good to clarify this in the spec.

**Comment by evilpie** [comment]
Do note that `'self'` blocking (not allowing) same-origin URLs is actually explicitly tested in https://wpt.fyi/results/content-security-policy/blob/blob-urls-do-not-match-self.sub.html?label=master&label=experimental&aligned&q=blob-urls-do-not-match-self.sub.html. So all browsers currently violate that part of the specification.

**Comment by antosart** [comment]
That's right. We should probably fix the spec and exclude the `blob` scheme explicitly when checking origin equality for the `'self'` keyword.

**Comment by annevk** [comment]
@antosart is excluding `blob:` URLs still the intent? WebKit currently follows the specification. (I'm not sure how that explains the test from @evilpie but perhaps we don't support `script-src-elem` and it's a false-positive?)

**Comment by antosart** [comment]
I think the rationale for differentiating blob URLs from normal same-origin network URLs is that the content of the blob URL comes from the page and is more susceptible to injection attacks. So it's safer if `self` does not allow blob: URLs.

I am curious about WebKit's behavior though, how do you explain the WPT test? (For script-src-elem it actually looks all [tests](https://wpt.fyi/results/content-security-policy/script-src-attr-elem?label=experimental&label=master&aligned) are passing, is that actually implemented?)

If all vendors were to agree already, and since, as I assume from how old the WPT above is, this has been the behavior from the beginning, I'd suggest we just fix the spec.

**Comment by annevk** [comment]
I'm not sure, maybe @roberto-apple can take a look. The code changes in https://github.com/WebKit/WebKit/pull/60814 give me the impression we are currently not excluding `blob:` URLs here.

Having said that, your rationale seems reasonable to me and we'd be okay with aligning on that, even if we're currently (maybe) not.

**Comment by roberto-apple** [comment]
Hello.
Regarding the WPT test `blob-urls-do-not-match-self.sub.html`, it looks like there's a simple gap in coverage in that it doesn't test frame-src

It currently tests this CSP
`<meta http-equiv="Content-Security-Policy" content="script-src 'self' 'unsafe-inline'; connect-src 'self'; child-src 'self';">`

But if we add this (and some additional test code, of course), the test would fail with WebKit's current baseline and pass with the proposed fix from the link.
`<meta http-equiv="Content-Security-Policy" content="script-src 'self' 'unsafe-inline'; connect-src 'self'; child-src 'self'; frame-src 'self';">`

WebKit currently has special handling for when the URL is a blob URL and the check is for frame-src where it extracts the blob's creator origin URL and matches that instead of the actual blob URL. The proposed code change in WebKit removes that. If the spec just needs to be updated and there is a gap in blob-urls-do-not-match-self.sub.html where frame-src should not match 'self', then we'll land the code change for WebKit.

**Comment by TimvdLippe** [comment]
Servo was failing those tests as well. #821 would be sufficient for Servo to pass the relevant tests.
